File: app/src/main/java/mods/net/Urban.java
package mods.net;

import androidx.fragment.app.FragmentActivity;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.URLEncoder;
import java.util.Random;

import mods.DiscordTools;
import mods.utils.ToastUtil;

public class Urban {

    public static void getDefinition(FragmentActivity context, String query) {
        DiscordTools.THREAD_POOL.execute(() -> {
            try {
                String input = query.replace("\n", "").trim();

                if (input.isEmpty()) {
                    ToastUtil.toast("Invalid query");
                    return;
                }

                String data = Net.nonAsyncRequest("https://api.urbandictionary.com/v0/define?term=" + URLEncoder.encode(input, "UTF-8"), null);

                if (data == null) {
                    ToastUtil.toast("Something went wrong, check your connection or search term and retry.");
                    return;
                }

                JSONArray list = new JSONObject(data).optJSONArray("list");

                if (list == null || list.length() == 0) {
                    ToastUtil.toast("'" + input + "' has no results.");
                    return;
                }

                final String result = "Result for " + input + ":\n\n" +
                        list.getJSONObject(new Random().nextInt(list.length()))
                                .optString("definition", "")
                                .replace("[", "")
                                .replace("]", "");

                context.runOnUiThread(() -> {
                    DiscordTools.newBuilder(context)
                            .setTitle("Urban Dictionary")
                            .setMessage(result)
                            .setNeutralButton("Copy", (dialog, which) -> {
                                DiscordTools.copyToClipboard(result);
                                ToastUtil.toast("Copied to clipboard");
                            })
                            .setPositiveButton("Exit", null)
                            .showSafely();
                });
            } catch (Exception e) {
                e.printStackTrace();
                ToastUtil.toast("Something went wrong.");
            }
        });
    }
}

File: app/src/main/java/mods/net/Urban.kt
package mods.net

import android.app.Activity
import mods.dialog.Dialogs
import mods.extensions.json
import mods.dialog.SimpleLoadingSpinner
import mods.promise.hideSpinner
import mods.utils.ClipboardUtil
import mods.utils.ToastUtil
import java.util.Random
import androidx.core.net.toUri

object Urban {

    @JvmStatic
    fun showDefinition(context: Activity, input: String) {
        val url = "https://api.urbandictionary.com/v0/define"
            .toUri()
            .buildUpon()
            .appendQueryParameter("term", input)
            .build()
            .toString()

        Net.doGetAsync(url).hideSpinner(
            SimpleLoadingSpinner(context).show("Loading definition...")
        ).subscribe({
            val list = it.json().optJSONArray("list")
            if (list == null || list.length() == 0) {
                ToastUtil.toast("'$input' has no results.")
                return@subscribe
            }
            val definitionText = list.getJSONObject(Random().nextInt(list.length()))
                .optString("definition", "")
                .replace("[", "")
                .replace("]", "")

            val result = "Result for $input:\n\n$definitionText"

            Dialogs.newBuilder(context)
                .setTitle("Urban Dictionary")
                .setMessage(result)
                .setNeutralButton("Copy") { _, _ ->
                    ClipboardUtil.copy(result, "Copied to clipboard")
                }
                .setPositiveButton("Exit")
                .showSafely()
        }, {
            ToastUtil.toast("Something went wrong, check your connection or search term and retry.")
        })
    }
}
